src/clientiface.cpp

```diff
@@ -168,24 +168,44 @@ void RemoteClient::GetNextBlocks (
 	// Get view range and camera fov (radians) from the client
 	s16 wanted_range = sao->getWantedRange() + 1;
 	float camera_fov = sao->getFov();
+	// in case the wanted_range was decreased by the client
+	if (m_blocks_occ.size() > wanted_range)
+		m_blocks_occ.resize(wanted_range);
 
 	/*
 		Get the starting value of the block finder radius.
 	*/
 	if (m_last_center != center) {
 		m_nearest_unsent_d = 0;
 		m_last_center = center;
+		// clear occlusion cache when we moved
+		m_blocks_occ.clear();
 	}
 	// reset the unsent distance if the view angle has changed more that 10% of the fov
 	// (this matches isBlockInSight which allows for an extra 10%)
 	if (camera_dir.dotProduct(m_last_camera_dir) < std::cos(camera_fov * 0.1f)) {
 		m_nearest_unsent_d = 0;
 		m_last_camera_dir = camera_dir;
+		// clear occlusion cache when we're looking somewhere else
+		m_blocks_occ.clear();
 	}
```

Comment:
Since everything gets lost if the player moves 16 blocks or his view by 10% I'm really wondering if the extra work (*) of distance-based caching is worth it.
Thoughts on returning to the initial, simpler implementation?

(*) and risk of there being a bug somewhere

Reply:
I hear you.

The simpler implementation was hard to reason about.
The server has two "modes":
1. If there is something to send it continuously loops over the layers (i.e. a particular distance) and sends blocks that are visible and the client has not seen. The same layer is usually visited multiple times, since there is a limit on how many blocks are sent per operation
2. If there is nothing to send, the server resets the minimum distance to 0 every 2s, and goes through all layers again.

To optimize (1) one just needs to cache occluded blocks for the current layer/distance, and then wipe it each time the layer changes. That part is easy.
Case (2) is more tricky. We need to make sure that the cache is correctly invalidated; that's mostly what this is doing. (2) is also very expensive, as it has to recalculate the occlusion status of all blocks again, just to see if it has changed and there is a potential block that needs to be sent. If we wipe the cache each time we won't have gained anything for this case.

This PR now only invalidates layers and it handles both cases correctly that could be affected, and, short of tracking the set of blocks each block is occluding, it is somewhat optimal.

The prior version where the cache was kept when nothing was sent in the last iteration was not correct. A block that was occluded could have changed, revealing other blocks.

Reply:
Also we'd get into state (1) above when we move or look around and (2) when the vantage point and direction have not changed a bit.

Actually... I can go back to the old implementation if I keep the part that removes a changed block from the occlusion list. In that case the server will send (at least) that block again (if it is now visible to the client), hence we have sent at least one block and the occlusion cache would be cleared. It would clear the entire cache, though, not just parts that are further away than a changed block.

Reply:
Sorry one more comment: After some more looking and testing it seems that case (2) cannot be optimized with caching. The whole logic between SetBlock(s)NotSent from AsyncRunStep with the timing in GetNextBlocks is just too obtuse and fragile. I'll likely give up on that part, and just to per layer caching that is cleared at the end of the layer. That takes care of a most of case (1).

Reply:
> Would you favor that? I'm actually not sure now what I would like better.

I think it's worth testing and seeing if that implementation has the same bugs you just mentioned in the other comment, if it does not then yes I'd prefer that.
IMO if anything, the old implementation is easier to reason about. It just:
* Never occlusion-checks any block twice
* Drops the cache if a block is modified *or* the camera moves